pgx/src/array.rs

```diff
@@ -232,20 +235,68 @@ impl ArrayPtr {
         // Most importantly, the caller has asserted this is in fact a valid [T],
         // by calling this function, so both this code and the caller can rely
         // on that assertion, only requiring that it is correct.
-        unsafe {
-            NonNull::new_unchecked(slice_from_raw_parts_mut(
-                pgx_ARR_DATA_PTR(self.at.as_ptr()).cast(),
-                len,
-            ))
-        }
+        unsafe { NonNull::new_unchecked(pgx_ARR_DATA_PTR(self.at.as_ptr()).cast()) }
     }
 }
 
 /**
 A pointer into a Postgres varlena.
 */
-pub struct RawArray<T> {
-    _ptr: ArrayPtr,
-    _len: usize,
-    ty: PhantomData<T>,
+pub struct RawArray {
+    ptr: ArrayPtr,
+    len: usize,
+}
+
+#[deny(unsafe_op_in_unsafe_fn)]
+impl RawArray {
+    /**
+    # Safety
+
+    Requires a pointer that really is from a properly constructed varlena,
+    using the format used internally by Postgres.
+    */
+    pub unsafe fn from_valid(ptr: ArrayPtr) -> RawArray {
+        let len = unsafe { ptr.len() } as usize;
+        RawArray { ptr, len }
+    }
+
+    fn len(&self) -> usize {
+        self.len
+    }
+
+    fn into_ptr(self) -> ArrayPtr {
+        self.ptr
+    }
+    
+    /// Oxidized form of ARR_NULLBITMAP(ArrayType*)
+    pub fn nulls(&self) -> Option<NonNull<[u8]>> {
+        // for expected behavior, see:
+        // postgres/src/include/utils/array.h
+        // #define ARR_NULLBITMAP
+        let len = self.len + 7 >> 3; // Obtains 0 if len was 0.
+
+        // SAFETY: This obtains the nulls pointer, which is valid to obtain because
+        // the validity was asserted on construction. However, unlike the other cases,
+        // it isn't correct to trust it. Instead, this gets null-checked.
+        // This is because, while the initial pointer is NonNull,
+        // ARR_NULLBITMAP can return a nullptr!
+        Some(unsafe { NonNull::new_unchecked(slice_from_raw_parts_mut(self.ptr.nulls()?.as_ptr(), len)) } )
+    }
+
+    /// # Safety
+    /// Asserts the data is mutable, properly aligned, all that jazz.
+    pub unsafe fn data<T>(&mut self) -> NonNull<[T]> {
```

Comment:
I don't think this needs to be unsafe if it loads the requirements into construction.

Reply:
Well, also, invalid pointers aren't unsafe to construct, yadda yadda.